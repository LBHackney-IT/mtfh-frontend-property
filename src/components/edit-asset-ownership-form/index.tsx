export * from "./edit-asset-ownership-form";
