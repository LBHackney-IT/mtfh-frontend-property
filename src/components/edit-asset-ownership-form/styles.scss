@import '~lbh-frontend/lbh/base';


#submit-edit-ownership-button {
    margin-right: 50px;
}

.edit-ownership-form-actions {

    button,
    a {
        width: 200px;
    }
}