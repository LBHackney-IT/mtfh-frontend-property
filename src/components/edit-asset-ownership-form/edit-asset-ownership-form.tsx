import React, { useState } from "react";
import { Link as RouterLink } from "react-router-dom";

import locale from "../../services/locale";

import { Center, Spinner } from "@mtfh/common";
import {
  Asset,
  PatchAssetLbhOwnershipRequest,
  patchAsset,
} from "@mtfh/common/lib/api/asset/v1";
import "./styles.scss";

interface EditAssetOwnershipFormProps {
  setShowSuccess: React.Dispatch<React.SetStateAction<boolean>>;
  setRequestError: React.Dispatch<React.SetStateAction<string | null>>;
  asset: Asset;
}

export const EditAssetOwnershipForm = ({
  setShowSuccess,
  setRequestError,
  asset,
}: EditAssetOwnershipFormProps) => {
  const [isCouncilProperty, setIsCouncilProperty] = useState<boolean>(
    asset.assetManagement?.isCouncilProperty,
  );
  const [formSubmitted, setFormSubmitted] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);

  const handleRadioButton = (value: boolean) => {
    setIsCouncilProperty(value);
  };

  const handleFormSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setLoading(true);

    const { assetManagement } = asset;

    const requestPayload: PatchAssetLbhOwnershipRequest = {
      assetManagement: {
        ...assetManagement,
        isCouncilProperty,
      },
    };

    patchAsset(asset.id, requestPayload, asset?.versionNumber?.toString() ?? "")
      .then(() => {
        setShowSuccess(true);
        setFormSubmitted(true);
      })
      .catch((err) => {
        console.error({ err });
        setRequestError(err.message);
      })
      .finally(() => {
        setLoading(false);
      });
  };

  return (
    <div>
      <form onSubmit={(e) => handleFormSubmit(e)}>
        <div className="govuk-form-group lbh-form-group">
          <fieldset className="govuk-fieldset" aria-describedby="edit-ownership-hint">
            <legend className="govuk-label lbh-label">
              Is this asset property of Hackney Council?
            </legend>
            <div className="govuk-radios lbh-radios">
              <div className="govuk-radios__item">
                <input
                  className="govuk-radios__input"
                  id="is-council-property-yes"
                  data-testid="is-council-property-yes"
                  name="is-council-property"
                  type="radio"
                  value="Yes"
                  checked={isCouncilProperty === true}
                  onChange={() => handleRadioButton(true)}
                  disabled={formSubmitted}
                />
                <label
                  className="govuk-label govuk-radios__label"
                  htmlFor="is-council-property-yes"
                >
                  <span className="govuk-heading-s govuk-!-margin-bottom-1">Yes</span>
                  {locale.lbhOwnershipInformation.editOwnershipFormYesDescription}
                </label>
              </div>
              <div className="govuk-radios__item">
                <input
                  className="govuk-radios__input"
                  id="is-council-property-no"
                  data-testid="is-council-property-no"
                  name="is-council-property"
                  type="radio"
                  value="No"
                  checked={isCouncilProperty === false}
                  onChange={() => handleRadioButton(false)}
                  disabled={formSubmitted}
                />
                <label
                  className="govuk-label govuk-radios__label"
                  htmlFor="is-council-property-no"
                >
                  <span className="govuk-heading-s govuk-!-margin-bottom-1">No</span>
                  {locale.lbhOwnershipInformation.editOwnershipFormNoDescription}
                </label>
              </div>
            </div>
          </fieldset>
        </div>
        {loading && (
          <Center>
            <Spinner />
          </Center>
        )}
        <div className="edit-ownership-form-actions">
          <button
            className="govuk-button lbh-button"
            data-module="govuk-button"
            type="submit"
            id="submit-edit-ownership-button"
            data-testid="submit-edit-ownership-button"
            disabled={
              formSubmitted ||
              isCouncilProperty === asset.assetManagement?.isCouncilProperty
            }
          >
            Confirm change
          </button>
          <RouterLink
            to={`/property/${asset.id}`}
            className="govuk-button govuk-secondary lbh-button lbh-button--secondary"
            id="cancel-edit-ownership-button"
            data-testid="cancel-edit-ownership-button"
          >
            Back to asset
          </RouterLink>
        </div>
      </form>
    </div>
  );
};
