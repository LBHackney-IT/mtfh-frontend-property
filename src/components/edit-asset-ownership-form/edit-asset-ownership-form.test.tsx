import React from "react";

import { render, server } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";
import { rest } from "msw";

import { EditAssetOwnershipForm } from "./edit-asset-ownership-form";

import { Asset } from "@mtfh/common/lib/api/asset/v1";
import userEvent from "@testing-library/user-event";

const testAsset: Asset = {
  id: "769894bd-b0bc-47eb-a780-322372c2448f",
  assetId: "0019062023",
  assetType: "Block",
  assetLocation: {
    parentAssets: [],
    floorNo: "0",
    totalBlockFloors: 3,
  },
  assetAddress: {
    uprn: "10008355547",
    addressLine1: "123 Test Block(CAN BE DELETED)",
    addressLine2: "",
    addressLine3: "",
    addressLine4: "",
    postPreamble: "",
    postCode: "FK8 1FH",
  },
  assetManagement: {
    isCouncilProperty: true,
    managingOrganisation: "London Borough of Hackney",
    managingOrganisationId: "c01e3146-e630-c2cd-e709-18ef57bf3724",
    isTMOManaged: false,
    agent: "Agent Smith",
    areaOfficeName: "Test",
    owner: "Test",
  },
  assetCharacteristics: {
    numberOfBedrooms: 0,
    numberOfLifts: 0,
    numberOfLivingRooms: 0,
    windowType: "Wood",
    yearConstructed: "1991",
    numberOfSingleBeds: null,
    numberOfDoubleBeds: null,
    numberOfFloors: null,
    totalBlockFloors: null,
    heating: null,
    propertyFactor: null,
    architecturalType: null,
  },
  versionNumber: 0,
  tenure: {
    id: "c01e3146-e630-c2cd-e709-18ef57bf3724",
    paymentReference: "",
    type: "",
    startOfTenureDate: "",
    endOfTenureDate: "",
    isActive: false,
  },
  rootAsset: "",
  parentAssetIds: "",
  rentGroup: null,
  boilerHouseId: "",
};

beforeEach(() => {
  jest.resetAllMocks();

  server.use(
    rest.patch(`/api/v1/assets/${testAsset.id}`, (req, res, ctx) => res(ctx.status(204))),
  );
});

// Form HTML elements
const submitButton = screen.getByRole("button", { name: /confirm change/i });
const radioOptionYes = screen.getByRole("radio", { name: /yes/i });
const radioOptionNo = screen.getByRole("radio", { name: /no/i });
const radioOptions = screen.getAllByRole("radio");

test("it renders the component", async () => {
  const { container } = render(
    <EditAssetOwnershipForm
      asset={testAsset}
      setShowSuccess={jest.fn()}
      setRequestError={jest.fn()}
    />,
  );

  expect(container).toMatchSnapshot();
});

test("it displays two radio buttons to set isCouncilProperty true or false", async () => {
  render(
    <EditAssetOwnershipForm
      asset={testAsset}
      setShowSuccess={jest.fn()}
      setRequestError={jest.fn()}
    />,
  );

  expect(radioOptions).toHaveLength(2);
  expect(radioOptionYes).toBeVisible();
  expect(radioOptionNo).toBeVisible();
});

test("the radio option is already checked on page load if isCouncilProperty has a value", async () => {
  testAsset.assetManagement.isCouncilProperty = true;

  render(
    <EditAssetOwnershipForm
      asset={testAsset}
      setShowSuccess={jest.fn()}
      setRequestError={jest.fn()}
    />,
  );

  expect(radioOptionYes).toBeChecked();
});

test("the submit button of the form starts out as disabled, and if the ownership of the asset is changed, the button is enabled", async () => {
  render(
    <EditAssetOwnershipForm
      asset={testAsset}
      setShowSuccess={jest.fn()}
      setRequestError={jest.fn()}
    />,
  );

  // Button is disabled to start with
  expect(submitButton).toBeDisabled();

  // Change ownership 
  userEvent.click(radioOptionNo);

  // Button is now enabled
  expect(submitButton).toBeEnabled();
});

test("once the form has been submitted, both radio options and the submit button are disabled", async () => {
  render(
    <EditAssetOwnershipForm
      asset={testAsset}
      setShowSuccess={jest.fn()}
      setRequestError={jest.fn()}
    />,
  );

  // Change ownership 
  userEvent.click(radioOptionNo);

  await waitFor(async () => {
    // Submit form
    userEvent.click(submitButton)

    // Delay to allow re-render
    await new Promise((resolve) => setTimeout(resolve, 100));

    // Submit button and the radio options are disabled
    expect(submitButton).toBeDisabled();
    expect(radioOptionYes).toBeDisabled();
    expect(radioOptionNo).toBeDisabled();
  });
});