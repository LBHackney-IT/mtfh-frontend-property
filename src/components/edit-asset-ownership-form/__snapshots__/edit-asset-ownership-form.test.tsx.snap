// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders the component 1`] = `
<div>
  <div>
    <form>
      <div
        class="govuk-form-group lbh-form-group"
      >
        <fieldset
          aria-describedby="edit-ownership-hint"
          class="govuk-fieldset"
        >
          <legend
            class="govuk-label lbh-label"
          >
            Is this asset property of Hackney Council?
          </legend>
          <div
            class="govuk-radios lbh-radios"
          >
            <div
              class="govuk-radios__item"
            >
              <input
                checked=""
                class="govuk-radios__input"
                data-testid="is-council-property-yes"
                id="is-council-property-yes"
                name="is-council-property"
                type="radio"
                value="Yes"
              />
              <label
                class="govuk-label govuk-radios__label"
                for="is-council-property-yes"
              >
                <span
                  class="govuk-heading-s govuk-!-margin-bottom-1"
                >
                  Yes
                </span>
                Including TMO properties and freehold/leasehold with service charges
              </label>
            </div>
            <div
              class="govuk-radios__item"
            >
              <input
                class="govuk-radios__input"
                data-testid="is-council-property-no"
                id="is-council-property-no"
                name="is-council-property"
                type="radio"
                value="No"
              />
              <label
                class="govuk-label govuk-radios__label"
                for="is-council-property-no"
              >
                <span
                  class="govuk-heading-s govuk-!-margin-bottom-1"
                >
                  No
                </span>
                Not owned by Hackney Council
              </label>
            </div>
          </div>
        </fieldset>
      </div>
      <div
        class="edit-ownership-form-actions"
      >
        <button
          class="govuk-button lbh-button"
          data-module="govuk-button"
          data-testid="submit-edit-ownership-button"
          disabled=""
          id="submit-edit-ownership-button"
          type="submit"
        >
          Confirm change
        </button>
        <a
          class="govuk-button govuk-secondary lbh-button lbh-button--secondary"
          data-testid="cancel-edit-ownership-button"
          href="/property/769894bd-b0bc-47eb-a780-322372c2448f"
          id="cancel-edit-ownership-button"
        >
          Back to asset
        </a>
      </div>
    </form>
  </div>
</div>
`;
